File: additionOfPolynomials.c
//Write a program to perform addition of two polynomials using linked list.

#include <stdio.h>
#include <stdlib.h>

// Node structure for polynomial term
struct Node {
    int coeff;
    int pow;
    struct Node* next;
};

// Create a new node
struct Node* createNode(int coeff, int pow) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->coeff = coeff;
    newNode->pow = pow;
    newNode->next = NULL;
    return newNode;
}

// Insert term at end
void insertTerm(struct Node** poly, int coeff, int pow) {
    struct Node* newNode = createNode(coeff, pow);
    if (*poly == NULL) {
        *poly = newNode;
    } else {
        struct Node* temp = *poly;
        while (temp->next != NULL)
            temp = temp->next;
        temp->next = newNode;
    }
}

// Display polynomial
void display(struct Node* poly) {
    while (poly != NULL) {
        printf("%dx^%d", poly->coeff, poly->pow);
        if (poly->next != NULL)
            printf(" + ");
        poly = poly->next;
    }
    printf("\n");
}

// Add two polynomials
struct Node* addPolynomials(struct Node* poly1, struct Node* poly2) {
    struct Node* result = NULL;

    while (poly1 != NULL && poly2 != NULL) {
        if (poly1->pow > poly2->pow) {
            insertTerm(&result, poly1->coeff, poly1->pow);
            poly1 = poly1->next;
        } else if (poly1->pow < poly2->pow) {
            insertTerm(&result, poly2->coeff, poly2->pow);
            poly2 = poly2->next;
        } else {
            insertTerm(&result, poly1->coeff + poly2->coeff, poly1->pow);
            poly1 = poly1->next;
            poly2 = poly2->next;
        }
    }

    // Remaining terms
    while (poly1 != NULL) {
        insertTerm(&result, poly1->coeff, poly1->pow);
        poly1 = poly1->next;
    }
    while (poly2 != NULL) {
        insertTerm(&result, poly2->coeff, poly2->pow);
        poly2 = poly2->next;
    }

    return result;
}

// Main
int main() {
    struct Node* poly1 = NULL;
    struct Node* poly2 = NULL;
    struct Node* result = NULL;

    // First polynomial: 3x^3 + 4x^2 + 5
    insertTerm(&poly1, 3, 3);
    insertTerm(&poly1, 4, 2);
    insertTerm(&poly1, 5, 0);

    // Second polynomial: 5x^2 + 6x + 7
    insertTerm(&poly2, 5, 2);
    insertTerm(&poly2, 6, 1);
    insertTerm(&poly2, 7, 0);

    printf("Polynomial 1: ");
    display(poly1);

    printf("Polynomial 2: ");
    display(poly2);

    result = addPolynomials(poly1, poly2);

    printf("Sum: ");
    display(result);

    return 0;
}

File: additionOfPolynomials.py
#Write a program to perform addition of two polynomials using linked list.

class Node:
    def __init__(self, coeff, pow):
        self.coeff = coeff
        self.pow = pow
        self.next = None

def insert_term(head, coeff, pow):
    new_node = Node(coeff, pow)
    if head is None:
        return new_node
    temp = head
    while temp.next:
        temp = temp.next
    temp.next = new_node
    return head

def display(poly):
    while poly:
        print(f"{poly.coeff}x^{poly.pow}", end="")
        if poly.next:
            print(" + ", end="")
        poly = poly.next
    print()

def add_polynomials(poly1, poly2):
    result = None
    while poly1 and poly2:
        if poly1.pow > poly2.pow:
            result = insert_term(result, poly1.coeff, poly1.pow)
            poly1 = poly1.next
        elif poly1.pow < poly2.pow:
            result = insert_term(result, poly2.coeff, poly2.pow)
            poly2 = poly2.next
        else:
            result = insert_term(result, poly1.coeff + poly2.coeff, poly1.pow)
            poly1 = poly1.next
            poly2 = poly2.next

    while poly1:
        result = insert_term(result, poly1.coeff, poly1.pow)
        poly1 = poly1.next
    while poly2:
        result = insert_term(result, poly2.coeff, poly2.pow)
        poly2 = poly2.next

    return result

# First Polynomial: 3x^3 + 4x^2 + 5
poly1 = None
poly1 = insert_term(poly1, 3, 3)
poly1 = insert_term(poly1, 4, 2)
poly1 = insert_term(poly1, 5, 0)

# Second Polynomial: 5x^2 + 6x + 7
poly2 = None
poly2 = insert_term(poly2, 5, 2)
poly2 = insert_term(poly2, 6, 1)
poly2 = insert_term(poly2, 7, 0)

print("Polynomial 1:")
display(poly1)
print("Polynomial 2:")
display(poly2)

result = add_polynomials(poly1, poly2)

print("Sum of polynomials:")
display(result)
